### DISCUSSION TITLE
ICYStream and WiFiManager - Initial WiFi credentials unknown ( Moved from wrong repository discussion)

### DISCUSSION THREAD

**Discussion by Ephesian** [pschatzmann/arduino-audio-tools#1761] [Q&A] [answered]
Thanks for your previous reply ->https://github.com/pschatzmann/arduino-audio-driver/discussions/38

I realised that I had posted in wrong place. I am very new to ESP32 and c++

I am using an AI Thinker AudioKit  V2.2 A247 with an E8388, and I can get a Sinewave output

The problem with this is that since I have moved the AudioTools objects into the code, the StreamCopy  object copier is not found 

**src/main.cpp: In function 'void loop()':
src/main.cpp:92:3: error: 'copier' was not declared in this scope
   copier.copy();
   ^~~~~~
src/main.cpp:92:3: note: suggested alternative: 'open'
   copier.copy();
   ^~~~~~
   open**

here is my sketch.

_/**
 * @brief We use a predefied board (AudioKitEs8388V1) and output_device a sine
 * with the help of the AudioTools I2SCodecStream
 * @author phil schatzmann
 */

#include <Arduino.h>
#include <SPI.h>
#include <WiFi.h> 

#include "AudioTools.h" // install https://github.com/pschatzmann/arduino-audio-tools
#include "AudioTools/AudioCodecs/CodecMP3Helix.h"
#include "AudioTools/AudioLibs/I2SCodecStream.h"

// Setup WiFi using WifiManager
#include <WiFiManager.h> // https://github.com/tzapu/WiFiManager

//AudioInfo info(44100, 2, 16);
// SineWaveGenerator<int16_t> sineWave(32000);
// GeneratedSoundStream<int16_t> sound(sineWave);

void setup() {
  // Setup logging
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);
  LOGLEVEL_AUDIODRIVER = AudioDriverDebug;
  WiFi.mode(WIFI_STA); // explicitly set mode, esp defaults to STA+AP 3010
 // WiFiManager, Local intialization. Once its business is done, there is no need to keep it around
  WiFiManager wm;

  // reset settings - wipe stored credentials for testing
  // these are stored by the esp library
  // wm.resetSettings();

  // Automatically connect using saved credentials,
  // if connection fails, it starts an access point with the specified name ( "AutoConnectAP"),
  // if empty will auto generate SSID, if password is blank it will be anonymous AP (wm.autoConnect())
  // then goes into a blocking loop awaiting configuration and will return success result

  bool res;
  // res = wm.autoConnect(); // auto generated AP name from chipid
  // res = wm.autoConnect("AutoConnectAP"); // anonymous ap
  res = wm.autoConnect("AutoConnectAP", "password"); // password protected ap

  if (!res)
  {
    Serial.println("Failed to connect");
    // ESP.restart();
  }
  else
  {
    // if you get here you have connected to the WiFi
    Serial.println("connected...yeey :)");
  }

  Serial.println(WiFi.SSID());
  Serial.println("");
  Serial.println("Wifi..."+WiFi.SSID()+"pwd"+ WiFi.psk());

String currentSSID = WiFi.SSID();
String currentPwd = WiFi.psk();
Serial.println(currentSSID+" ..."+currentPwd);

const char* ssid = currentSSID.c_str();
const char* pwd = currentPwd.c_str();

ICYStream url(ssid,pwd);

I2SCodecStream out(AudioKitEs8388V1);
EncodedAudioStream dec(&out, new MP3DecoderHelix()); // Decoding stream
//StreamCopy copier(out, sound);
StreamCopy copier(dec, url);
  
  // start I2S & codec with i2c and i2s configured above
  Serial.println("starting I2S...");
  //auto config = out.defaultConfig();
  auto config = out.defaultConfig();
  
  out.begin(config);
  out.setVolume(.5);
  dec.begin();
  url.begin("http://31.132.6.142:4231/stream?type=http&nocache=7361","audio/mp3");
  //url.begin("http://stream.srg-ssr.ch/m/rsj/mp3_128","audio/mp3");
Serial.println(WiFi.localIP());
// Setup sine wave
//sineWave.begin(info, N_B4);
}

// Arduino loop - copy sound to out
void loop() { 
  copier.copy();
}_

**Comment by pschatzmann** [comment, marked as answer]
This is impossible to read: Please use the proper [code formatting syntax](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code),

Dont define variables locally that you want to be available across your sketch.

Maybe you should start with some C++ tutorial first...
